File: src/skills/memory.ts
// src/skills/memory.ts
//
// Cross-session memory tool for the AI agent. Lets the model remember long-term
// user preferences and configuration and recall them in later chats. Backed by a
// flat key/value store in ~/.jedi/agent_memory.json via the Rust `memory_*`
// commands. Only touches Jedi's own private storage, so it is friction-free.

import { invoke } from '@tauri-apps/api/core'
import type { ParameterSchema, Skill } from './types'

const memoryParameters: ParameterSchema = {
  type: 'object',
  properties: {
    action: {
      type: 'string',
      description: "Operation to perform: 'save', 'recall', 'list', or 'delete'",
      required: true,
    },
    key: {
      type: 'string',
      description: 'Memory key, e.g. preferred_wallpaper. Required for save / recall / delete.',
      required: false,
    },
    value: {
      type: 'string',
      description: 'Memory content to store. Required for save.',
      required: false,
    },
  },
  required: ['action'],
}

interface MemoryEntry {
  key: string
  value: string
}

async function executeMemory(args: {
  action: string
  key?: string
  value?: string
}): Promise<string> {
  const action = (args.action || '').toLowerCase()
  switch (action) {
    case 'save': {
      if (!args.key || args.value === undefined) return '错误：save 需要 key 和 value'
      return invoke<string>('memory_save', { key: args.key, value: args.value })
    }
    case 'recall': {
      if (!args.key) return '错误：recall 需要 key'
      return invoke<string>('memory_recall', { key: args.key })
    }
    case 'list': {
      const entries = await invoke<MemoryEntry[]>('memory_list')
      if (!entries.length) return '（暂无记忆）'
      return entries.map(e => `${e.key}: ${e.value}`).join('\n')
    }
    case 'delete': {
      if (!args.key) return '错误：delete 需要 key'
      return invoke<string>('memory_delete', { key: args.key })
    }
    default:
      return `未知操作：${args.action}（应为 save / recall / list / delete）`
  }
}

export const memorySkill: Skill = {
  id: 'memory',
  name: 'MEMORY',
  description:
    'Remember long-term user preferences and configuration across chat sessions. ' +
    'action=save stores a key/value pair; action=recall retrieves a value by key; ' +
    'action=list shows everything remembered; action=delete removes a key. ' +
    'Use this whenever the user asks you to remember something for later.',
  icon: '🧠',
  enabled: true,
  autoCallable: true,
  parameters: memoryParameters,
  execute: executeMemory,
}

File: src-tauri/src/api/memory.rs
// Cross-session memory for the AI agent.
//
// Lets the agent remember long-term user preferences and frequently-used
// configuration across chat sessions. Persisted as a flat key/value map in
// ~/.jedi/agent_memory.json. This only touches Jedi's own private storage
// (never the system), so it is a low-risk, friction-free "remember / recall".

use serde::Serialize;
use std::collections::BTreeMap;
use std::path::PathBuf;

/// 记忆文件路径（~/.jedi/agent_memory.json），必要时创建目录
fn memory_path() -> Result<PathBuf, String> {
  let home = dirs::home_dir().ok_or_else(|| "无法获取用户主目录".to_string())?;
  let dir = home.join(".jedi");
  if !dir.exists() {
    std::fs::create_dir_all(&dir).map_err(|e| format!("创建 .jedi 目录失败: {}", e))?;
  }
  Ok(dir.join("agent_memory.json"))
}

fn load() -> Result<BTreeMap<String, String>, String> {
  let path = memory_path()?;
  Ok(
    std::fs::read_to_string(&path)
      .ok()
      .and_then(|s| serde_json::from_str(&s).ok())
      .unwrap_or_default(),
  )
}

fn store(map: &BTreeMap<String, String>) -> Result<(), String> {
  let path = memory_path()?;
  let s = serde_json::to_string_pretty(map).map_err(|e| e.to_string())?;
  std::fs::write(&path, s).map_err(|e| format!("写入记忆失败: {}", e))
}

#[derive(Debug, Serialize)]
pub struct MemoryEntry {
  pub key: String,
  pub value: String,
}

/// 记住一条长期信息（用户偏好、常用配置等），跨会话保留
#[tauri::command]
pub fn memory_save(key: String, value: String) -> Result<String, String> {
  let mut map = load()?;
  map.insert(key.clone(), value);
  store(&map)?;
  Ok(format!("已记住「{}」", key))
}

/// 按键回忆一条此前记住的信息
#[tauri::command]
pub fn memory_recall(key: String) -> Result<String, String> {
  let map = load()?;
  Ok(match map.get(&key) {
    Some(v) => v.clone(),
    None => format!("未找到记忆「{}」", key),
  })
}

/// 列出所有已记住的信息
#[tauri::command]
pub fn memory_list() -> Result<Vec<MemoryEntry>, String> {
  let map = load()?;
  Ok(
    map
      .into_iter()
      .map(|(key, value)| MemoryEntry { key, value })
      .collect(),
  )
}

/// 删除一条记忆
#[tauri::command]
pub fn memory_delete(key: String) -> Result<String, String> {
  let mut map = load()?;
  if map.remove(&key).is_some() {
    store(&map)?;
    Ok(format!("已删除记忆「{}」", key))
  } else {
    Ok(format!("未找到记忆「{}」", key))
  }
}
